libresin/libresin/core/resources/shader_resource.cpp

```diff
@@ -0,0 +1,202 @@
+    return;
+  }
+
+  ext_defi_contents_[*it] = std::move(defi_content);
+  is_dirty_               = true;
+}
+
+bool ShaderResource::is_glsl_ready() const { return ext_defi_contents_.size() == ext_defi_names_.size(); }
+
+const std::string& ShaderResource::get_raw() const { return raw_content_; }
+
+const std::string& ShaderResource::get_glsl() const {
+  if (!is_dirty_) {
+    return glsl_;
+  }
+
+  glsl_.clear();
+  glsl_.append(shader_macros::kSupportedShaderVersion);
+  for (const auto& ext_defi : ext_defi_contents_) {
+    glsl_.append(std::format("#define {} {}\n", ext_defi.first, ext_defi.second));
+  }
+  glsl_.append(raw_content_);
+
+  is_dirty_ = false;
+  return glsl_;
+}
+
+static ShaderType get_sh_type(const std::filesystem::path& path) {
+  auto file_ext = path.extension().string();
+  auto sh_type  = extension_to_shader_type(file_ext);
+  if (!sh_type.has_value()) {
+    helpers::log_file_err(path, "File extension \"{}\" is not suported.", file_ext);
+
+    throw FileExtensionNotSupportedException();
+  }
+
+  return sh_type.value();
+}
+
+static std::string load_content(const std::filesystem::path& path) {
+  namespace fs = std::filesystem;
+
+  if (!fs::exists(path)) {
+    helpers::log_file_err(path, "File does not exist.");
+    throw FileDoesNotExistException();
+  }
+
+  if (!fs::is_regular_file(path) && !fs::is_symlink(path)) {
+    helpers::log_file_err(path, "File must be a regular file or symlink.");
+    throw InvalidFileTypeException();
+  }
+
+  std::ifstream file_stream(path.string());
+  if (!file_stream.is_open()) {
+    helpers::log_file_err(path, "File stream cannot be opened.");
+    throw FileStreamNotAvailableException();
+  }
+
+  std::stringstream buffer;
+  buffer << file_stream.rdbuf();
+  return buffer.str();
+}
+
+static std::optional<std::filesystem::path> process_include_macro(std::string_view arg, int64_t curr_line_num) {
+  if (!arg.starts_with("\"") || !arg.ends_with("\"") || arg.size() < 2) {
+    helpers::log_sh_err(curr_line_num, "The include macro argument should begin and end with `\"`.");
+    return std::nullopt;
+  }
+
+  auto arg_val = std::string_view{arg.substr(1, arg.size() - 2)};
+  auto path    = std::filesystem::path{arg_val};
+  if (path.empty()) {
+    helpers::log_sh_err(curr_line_num, "The include macro argument cannot be empty.");
+    return std::nullopt;
+  }
+
+  if (path.is_absolute()) {
+    helpers::log_sh_err(curr_line_num, "The include macro argument cannot be an absolute path.");
+    return std::nullopt;
+  }
+  auto dep_ext = resin::extension_to_shader_type(path.extension().string());
+
+  if (!dep_ext.has_value() || dep_ext.value() != ShaderType::Library) {
+    helpers::log_sh_err(curr_line_num, "The include macro argument must a library shader (.glsl extension).");
+    return std::nullopt;
+  }
+
+  return path;
+}
+
+static std::optional<std::string> process_ext_defi_macro(std::string_view arg, int64_t curr_line_num) {
```

Comment:
```suggestion
static std::optional<std::string> process_ext_defi_macro_arg(std::string_view arg, int64_t curr_line_num) {
```
and `process_ext_defi_macro` method handling whole macro

Reply:
✅